Cilium operator HA mode

fristonio opened the pull request:
Initial implementation of HA mode cilium operator using Kubernetes leader election library.

![image](https://user-images.githubusercontent.com/21292343/86510214-6ebcb500-be0b-11ea-8e39-d93360ebe3ed.png)

After deleting the pod which is currently leading, the other operator pod wins the leader election and starts performing operations.

![image](https://user-images.githubusercontent.com/21292343/86510302-2ce03e80-be0c-11ea-9d0d-ec9bc4d17ef8.png)

Fixes #12330

sayboras commented on operator/main.go:
```diff
@@ -148,6 +159,79 @@ func runOperator(cmd *cobra.Command) {
+	hostID = rand.RandomStringWithPrefix(hostID+"-", 8)
+	log.Infof("using host ID %s for operator POD", hostID)
+
+	ns := option.K8sNamespaceName
```
will it be `ns := viper.GetString(option.K8sNamespaceName)` instead ?

aanm replied:
@sayboras no, should be `option.Config.K8sNamespace` as it is set in https://github.com/cilium/cilium/blob/edff374340cb55318ea3fffba13322f9152cdaf6/pkg/option/config.go#L2535

aanm commented on install/kubernetes/cilium/charts/operator/templates/clusterrole.yaml:
```diff
@@ -57,3 +57,11 @@ rules:
+  # to obtain lease lock for leader election.
+  - leases
+  verbs:
+  - '*'
```
needs a newline here

aanm commented on operator/main.go:
```diff
@@ -148,6 +157,77 @@ func runOperator(cmd *cobra.Command) {
+	hostID = rand.RandomStringWithPrefix(hostID+"-", 8)
+	log.Infof("using host ID %s for operator POD", hostID)
+
+	ns := os.Getenv("NAMESPACE")
```
There's an environment label already with the namespace `CILIUM_K8S_NAMESPACE`

fristonio replied:
This is the outdated file, I am using `option.K8sNamespaceName` instead of the environment variable now.

sayboras replied:
I think the comments got collapsed. As mentioned by @aanm https://github.com/cilium/cilium/pull/12409#discussion_r450082036, it should be `option.Config.K8sNamespace` right ( as `option.K8sNamespaceName` is _const_ ) ?

fristonio replied:
Ah, I see. Sorry, missed this while fixing the commits.

aanm commented on install/kubernetes/cilium/charts/operator/templates/clusterrole.yaml:
```diff
@@ -57,3 +57,11 @@ rules:
   - get
   - list
   - watch
+- apiGroups:
```
I guess this change should be part of the next commit?

aanm commented on install/kubernetes/cilium/charts/operator/templates/deployment.yaml:
```diff
@@ -7,7 +7,7 @@ metadata:
   namespace: {{ .Release.Namespace }}
 spec:
-  replicas: 1
+  replicas: 2
```
I guess this change should be part of the next commit?

sayboras commented on operator/main.go:
```diff
@@ -148,6 +159,79 @@ func runOperator(cmd *cobra.Command) {
+
+	// Get hostname for identity name of the lease lock holder.
+	// We identify the leader of the operator cluster using hostname.
+	hostID, err := os.Hostname()
```
sorry to chime in, I am planning to do something similar at work, so just take this chance to understand more.

Seems like any random id is fine, right ? Any particular reason to have hostName prefix?

fristonio replied:
Just to have some sort of context for the leaseholder ID.

sayboras commented on operator/main.go:
```diff
@@ -148,6 +159,79 @@ func runOperator(cmd *cobra.Command) {
+		Callbacks: leaderelection.LeaderCallbacks{
+			OnStartedLeading: onOperatorStartLeading,
+			OnStoppedLeading: func() {
+				log.Infof("leader election lost: %s", hostID)
```
nits: do you want to keep it consistent for logging, something like log.WithField("lease_id", hostID).Info("leader election lost") ?

fristonio replied:
Yeah, I think it makes sense. I will update the logging statements.

sayboras commented on operator/main.go:
```diff
@@ -148,6 +159,79 @@ func runOperator(cmd *cobra.Command) {
+				if identity == hostID {
+					log.Info("leading the operator HA deployment")
+				} else {
+					log.Infof("new leader elected: %s", identity)
```
same as above, log.WithField() can be used for more structured logging.

aanm commented on Documentation/cmdref/cilium-operator-aws.md:
```diff
@@ -15,52 +15,55 @@ cilium-operator-aws [flags]
-      --unmanaged-pod-watcher-interval int      Interval to check for unmanaged kube-dns pods (0 to disable) (default 15)
-      --update-ec2-apdater-limit-via-api        Use the EC2 API to update the instance type to adapter limits
-      --version                                 Print version information
+      --aws-instance-limit-mapping map            Add or overwrite mappings of AWS instance limit in the form of {"AWS instance type": "Maximum Network Interfaces","IPv4 Addresses per Interface","IPv6 Addresses per Interface"}. cli example: --aws-instance-limit-mapping=a1.medium=2,4,4 --aws-instance-limit-mapping=a2.somecustomflavor=4,5,6 configmap example: {"a1.medium": "2,4,4", "a2.somecustomfl ...
```
This commit could be squashed with a previous one.

fristonio replied:
Will do that.

aanm reviewed (changes requested):
:100:

sayboras commented on operator/main.go:
```diff
@@ -148,6 +159,78 @@ func runOperator(cmd *cobra.Command) {
+
+	// Use a Go context so we can tell the leaderelection code when we
+	// want to step down
+	ctx, contextCancelFunc := context.WithCancel(context.Background())
```
Will this contextCancelFunc variable be shadowed (i.e. it's new local variable instead of package level variable) ?

sayboras replied:
I just quickly try in playground https://play.golang.org/p/qSPp12Ps0XW

fristonio replied:
Yeah, you are right :sweat_smile: . I will push the required changes.

sayboras commented on operator/main.go:
```diff
@@ -96,6 +103,7 @@ func main() {
 		<-signals
 		gops.Close()
 		close(shutdownSignal)
+		contextCancelFunc()
```
Related the the comment ~below~ above, this package contextCancelFunc variable might be nil.

christarazi commented on operator/main.go:
```diff
@@ -148,6 +161,74 @@ func runOperator(cmd *cobra.Command) {
+	// We identify the leader of the operator cluster using hostname.
+	operatorID, err := os.Hostname()
+	if err != nil {
+		log.WithError(err).Fatal("failed to get hostname")
```
Nit: this msg might be cryptic when looking through logs when debugging; it's a bit too generic. Maybe something like:

 ```suggestion
		log.WithError(err).Fatal("Failed to get hostname when generating lease lock identity")
```

christarazi commented on operator/main.go:
```diff
@@ -148,6 +161,74 @@ func runOperator(cmd *cobra.Command) {
+
+	ns := option.Config.K8sNamespace
+	// If due to any reason the NAMESPACE is not set we assume it to be
+	// in default namespace.
```
Nit: previously, it looks like you were using the "NAMESPACE" env var, and this comment is a hold-over?

fristonio replied:
Yeah, I think I should update it to `CILIUM_K8S_NAMESPACE` as this is the environment variable which populates the option config for K8sNamespace.

christarazi commented on operator/main.go:
```diff
@@ -148,6 +161,74 @@ func runOperator(cmd *cobra.Command) {
+		Callbacks: leaderelection.LeaderCallbacks{
+			OnStartedLeading: onOperatorStartLeading,
+			OnStoppedLeading: func() {
+				log.WithField("operator_id", operatorID).Info("leader election lost")
```
Nit: do you mind capitalizing the first word of each log msg? This helps keep the log formatting consistent across the codebase. (While I don't think every log msg in the codebase now starts with a capital, we should converge on one style for the sake of readability when reading logs. :) )

fristonio replied:
Yeah, I think it makes sense to have this consistency. I will update the PR and will make sure this doesn't happen next time.

christarazi commented on operator/main.go:
```diff
@@ -148,6 +161,74 @@ func runOperator(cmd *cobra.Command) {
+				gops.Close()
+				close(shutdownSignal)
+				leaderElectionCtxCancel()
+				os.Exit(0)
```
More of a question here: is the `Exit` in order for the previous cilium-operator leader to become a "follower" (non-leader)?

fristonio replied:
I don't think if I understand the question correctly. :sweat_smile: 

The reason why we do an exit here is that once an operator stops leading, there is a lot of complexity in terms of stopping all the watchers gracefully managed by the operator. A more general approach I have seen with leader election in operators is to exit when once an already leading instance stops leading. Like here in tidb-operator - https://github.com/pingcap/tidb-operator/blob/master/cmd/controller-manager/main.go#L237-L239

ungureanuvladvictor replied:
I took a look at what kcm/scheduler do and both when they stop leading call klog.Fatalf and exit. 


I suspect this is going to cause some issues for the operator at least in the AWS ENI mode.

Think of the following case:
- user runs operator and configures the CiliumNode to set DeleteOnTermination for the ENIs that get created
- CN comes online and operator creates the ENI in AWS
- operator loses the leadership but it did not get a change to attach the ENI and set delete on termination
- user is left with a dangling ENI in the AWS account

I don't think we should over-engineer a solution now but should be aware of these edge cases in the future.

aanm replied:
That's a good point. Cilium can also crash in the middle of that operation. Is it possible to know which ENIs are dangling? I'm just thinking on the case where we add a GC to the operator to clean up such dangling ENIs.

ungureanuvladvictor replied:
Yep -- we can do it based on the tags (or names) of the ENIs. I think this exposes the fact that the operator is doing a lot of things that are not atomic and we do not have a good way to cancel/repair those operations when we lose leadership.

christarazi replied:
This is a good discussion and I've opened an issue to track these things. Please add whatever comes to mind in issue https://github.com/cilium/cilium/issues/12448

christarazi replied:
Essentially I was asking if the `Exit` is for transitioning between leader to non-leader, and it looks like the answer is yes. @fristonio

christarazi reviewed (approved):
Well done! Code looks good, just have a few minor comments.

aanm commented on operator/main.go:
```diff
@@ -148,6 +161,74 @@ func runOperator(cmd *cobra.Command) {
+				if identity == operatorID {
+					log.Info("Leading the operator HA deployment")
+				} else {
+					log.WithField("operator_id", operatorID).Infof("Operator with ID %s elected as new leader", identity)
```
`s/operator_id/operator-id`

aanm commented on install/kubernetes/cilium/charts/operator/templates/clusterrole.yaml:
```diff
@@ -65,3 +66,17 @@ rules:
+  - create
+  - get
+  - update
+{{- end }}
```
needs a new line

aanm replied:
you can put an `{{ else }}` here and use everything from line 60 to 68 in this else statement

aanm reviewed (changes requested):
The operator needs to detect that the k8s version has changed. For example if the k8s version running is 1.15 and if the operator is using the configmaplock, the operator should stop its execution.

christarazi commented on install/kubernetes/cilium/charts/operator/templates/clusterrole.yaml:
```diff
@@ -57,3 +57,26 @@ rules:
+  verbs:
+  - '*'
+{{- end }}
+{{- if and (lt .Capabilities.KubeVersion.Minor "14") (eq .Capabilities.KubeVersion.Major "1") }}
```
I think it would be helpful to add a comment for why we are checking for that specific K8s version.

fristonio replied:
Will do.

christarazi commented on install/kubernetes/experimental-install.yaml:
```diff
@@ -351,6 +351,14 @@ rules:
+  # to obtain lease lock for leader election.
+  - leases
+  verbs:
+  - '*'
```
Will we need the same check as above on the K8s version?

fristonio replied:
These are autogenerated files using the helm templates. I don't think we can add checks to these files, as they will be overridden next time we autogenerate install manifests.

christarazi commented on install/kubernetes/quick-install.yaml:
```diff
@@ -280,6 +280,14 @@ rules:
+  # to obtain lease lock for leader election.
+  - leases
+  verbs:
+  - '*'
```
And here as well, we might need the same check as above on the K8s version?

aanm replied:
@christarazi these are auto generated, given that the last version of k8s that requires the configmap based lease is k8s 1.14, which is deprecated for months now, I think it's fare to leave it as it is and not include the rules to access the configmap in the `quick-install.yaml`.

joestringer replied:
I agree, if there is any aspect here that involves additional extra work to support this operator HA feature on k8s 1.14 or below then I don't think it's worth adding that support to Cilium at this point.

aanm commented on pkg/k8s/version/version.go:
```diff
@@ -119,14 +125,29 @@ func updateServerGroupsAndResources(apiGroups []*metav1.APIGroup, apiResourceLis
+
+		if cached.capabilities.LeasesResourceLock && cached.capabilities.EndpointSlice {
+			break
+		}
```
please add a comment in this line. To avoid mistakes I think we  should not even add this. In case in the future we add a new capability and forget to update this line.

aanm commented on pkg/k8s/version/version.go:
```diff
@@ -164,6 +185,72 @@ func fallbackDiscovery(client kubernetes.Interface) error {
+	// ClusterRole did not had access to corresponding API resource.
+	// Look at the helm manifest to check the ClusterRole associated with cilium Deployments
+	// corresponding to the configured Kubernetes version.
+	if cached.version.Minor < 14 && cached.version.Major == 1 {
```
Instead of doing this, create a variable similar to the ones in line 80-82 of this file

aanm commented on pkg/k8s/version/version.go:
```diff
@@ -164,6 +185,72 @@ func fallbackDiscovery(client kubernetes.Interface) error {
+		return nil
+	}
+
+	_, err := client.CoordinationV1().Leases("kube-system").Get(context.TODO(), "kube-controller-manager", metav1.GetOptions{})
```
Add a comment stating what this is doing.

aanm commented on pkg/k8s/version/version.go:
```diff
@@ -164,6 +185,72 @@ func fallbackDiscovery(client kubernetes.Interface) error {
+		return nil
+	}
+
+	switch t := err.(type) {
```
There is a k8s helper function to check for error not found.
https://github.com/cilium/cilium/blob/f694b3ba28ac2936c413fddec7771ef14a2c8808/vendor/k8s.io/apimachinery/pkg/api/errors/errors.go#L485-L488

sayboras commented on test/kubernetes-test.sh:
```diff
@@ -6,7 +6,7 @@
 # process the test will fail.
 
-helm template install/kubernetes/cilium \
+helm install cilium install/kubernetes/cilium \
```
I assume the changes in this file is to overcome the issue, which _--kube-version_ flag is not available in helm 3 template CLI. Nice :tada: 

One minor nit, the current script is kind of idempotent (e.g. you can re-run a few times for same cluster), however _helm install_ is not. If you want, you can make it to `helm upgrade -i` (I used this command in CI a lot for same use case)

> helm upgrade -h
> ...
>   -i, --install                      if a release by this name doesn't already exist, run an install

fristonio replied:
Hey @sayboras 
Yeah, this is to overcome the default faked value provided for Built-In objects like `Capabilities.KubeVersion` when generating templates. I was just testing it out though. I will see the use of the command you mentioned. But, it seems to me that this might not be fixed by just doing a `helm install` or `helm upgrade` here. In the ginkgo test suite, we redeploy cilium multiple times, each time it makes use of this template generated earlier using the `helm template` command. See [here](https://github.com/cilium/cilium/blob/master/test/helpers/kubectl.go#L2316-L2345)
I have already asked a question related to this on slack and I will replace the use of generated templates for deploying cilium with running `helm upgrade -i` if everyone agrees.

aanm commented on install/kubernetes/cilium/charts/operator/templates/clusterrole.yaml:
```diff
@@ -57,11 +57,31 @@ rules:
   - get
   - list
   - watch
+# For cilium-operator running in HA mode.
```
I think we decided to drop the support for the `ConfigMapLock` and not support cilium operator HA in k8s <= 1.14?

fristonio committed:
vendor: vendor kubernetes leaderelection library

Signed-off-by: Deepesh Pathak <deepshpathak@gmail.com>

fristonio committed:
k8s: add coordinationv1 capability check to k8s version package

* Introduces config option `K8sLeasesFallbackDiscoveryEnabled` to check
  if fallback discovery is enabled for Leases.
* K8sLeasesFallbackDiscovery is enabled by default only in
  operator.

Signed-off-by: Deepesh Pathak <deepshpathak@gmail.com>

fristonio committed:
operator: support HA mode for operator using k8s leaderelection library

* Make leaderelection parameters configurable using command line flags
* Update cmdref to include documentation for new flags.

Signed-off-by: Deepesh Pathak <deepshpathak@gmail.com>

fristonio committed:
install: update helm templates to add HA capabilities for operator

Signed-off-by: Deepesh Pathak <deepshpathak@gmail.com>

fristonio committed:
test: generate cilium helm template validating against k8s cluster

* Use --validate with `helm template` command to validate the generated
  manifest against the associated kubernetes cluster
* For more information see -
  https://github.com/cilium/cilium/pull/12409#discussion_r453313631

Signed-off-by: Deepesh Pathak <deepshpathak@gmail.com>

aanm commented on pkg/k8s/version/version.go:
```diff
@@ -137,19 +162,92 @@ func fallbackDiscovery(client kubernetes.Interface) error {
+
+	// Unknown error, we can't derive whether to enable or disable
+	// EndpointSlices and need to error out.
+	return fmt.Errorf("Unable to validate EndpointSlices support: %s", err)
```
`s/Unable/unable`

aanm reviewed (approved):
Just a small nit, does not require my 2nd review.

christarazi commented on pkg/k8s/version/version.go:
```diff
@@ -49,6 +48,10 @@ type ServerCapabilities struct {
 	EndpointSlice bool
+
+	// LeasesResourceLock is the ability of K8S server to support Lease type
+	// from CoordinationV1 API for leader election purposes.
```
Nit: 

```suggestion
	// LeasesResourceLock is the ability of K8s server to support Lease type
	// from CoordinationV1 API for leader election purposes.
```

Could you also add a reference to the K8s docs in this comment? I've [mentioned previously in another PR](https://github.com/cilium/cilium/pull/12611#pullrequestreview-455054054) that we should more thoroughly document when we make changes to the K8s version / capabilities related code.

At a high-level it would be nice to have:

- A reference link to K8s docs so that our future selves can understand what this thing is and potentially why we need it.
- The relevant versions this capability maps to, so that when we remove / add support for certain K8s versions, we are able to know if this applicability is relevant or not.

christarazi replied:
To add another reason why I think it's important to document these things, is because we rarely make changes to this code / package[1]. It averages to around once every ~3 months, so lots of time / opportunity to lose context.

[1]: `git log --oneline --format=format:"%cs %s" pkg/k8s/version`

fristonio replied:
Sounds good, I will refactor the documentation.

christarazi commented on pkg/k8s/version/version.go:
```diff
@@ -137,21 +162,94 @@ func fallbackDiscovery(client kubernetes.Interface) error {
+
+	cached.mutex.RLock()
+	// Here we check if cilium has required authorization to check for leases object or not.
+	// Prior to v1.14 leases.coordination.k8s.io/v1 was not supported and hence cilium
```
I see a reference to document the version so that's good :+1: , but it would be nice to centralize it.

christarazi commented on pkg/k8s/version/version.go:
```diff
@@ -137,21 +162,94 @@ func fallbackDiscovery(client kubernetes.Interface) error {
+	}
+
+	if err != nil {
+		return fmt.Errorf("Cannot parse k8s server version from %+v: %s", sv, err)
```
Could you please ensure that all errors are formatted starting with a lower-case?

christarazi commented on pkg/option/config.go:
```diff
@@ -2044,6 +2054,19 @@ func (c *DaemonConfig) K8sAPIDiscoveryEnabled() bool {
+}
+
+// EnableK8sLeasesFallbackDiscovery enables using direct API probing as a fallback to check
+// for the support of Leases when discovering API groups is not possible
```
Nit: 

```suggestion
// for the support of Leases when discovering API groups is not possible.
```

(missing period at end)

christarazi commented on operator/main.go:
```diff
@@ -161,6 +181,85 @@ func runOperator(cmd *cobra.Command) {
+	// We only support operator in HA mode for Kubernetes Versions having support for
+	// LeasesResourceLock
+	// LeasesResourceLock type is preferred for resourcelocks because edits to Leases are less common
+	// and fewer objects in the cluster watch "all Leases".
```
```suggestion
	// If the K8s version has support for coordination.k8s.io/v1 use LeasesResourceLock
	// for resourcelock config else fallback to non HA mode for Operator.
	// We only support Operator in HA mode for Kubernetes Versions having support for
	// LeasesResourceLock.
	// LeasesResourceLock type is preferred for resourcelocks because edits to Leases are less common
	// and fewer objects in the cluster watch "all Leases".
```

I think the majority of this comment can be folded into the main documentation for the capability itself. We can keep a short explanation of why we're doing this check and then something like "See docs on `capabilities.LeasesResourceLock` for more context."

fristonio replied:
Makes sense, I will update the documentation.

christarazi commented on install/kubernetes/cilium/charts/operator/templates/clusterrole.yaml:
```diff
@@ -57,3 +57,22 @@ rules:
+# Cilium operator running in HA mode requires the use of ResourceLock for Leader Election
+# between mulitple running instances.
+# The preferred way of doing this is to use LeasesResourceLock as edits to Leases are less
+# common and fewer objects in the cluster watch "all Leases".
+# The support for leases was introduced in coordination.k8s.io/v1 during Kubernetes 1.14 release.
+# In Cilium we currently don't support HA mode for K8s version < 1.14. This condition make sure
```
This is good. Just pointing it out here in case you may need to make this consistent with the other docs changes.

christarazi commented on install/kubernetes/cilium/charts/operator/templates/deployment.yaml:
```diff
@@ -7,7 +7,12 @@ metadata:
   name: cilium-operator
   namespace: {{ .Release.Namespace }}
 spec:
+  # We support HA mode only for Kubernetes version > 1.14
```
It'd be nice to a have a reference to give more context here as to why. Something like "See \<file or something\> for more details."

christarazi commented on test/helpers/kubectl.go:
```diff
@@ -3833,14 +3837,19 @@ func (kub *Kubectl) HelmAddCiliumRepo() *CmdRes {
+// .Capabilities.KubeVersion with the values from associated cluster.
+// This comes with a caveat that the command might fail if helm is not able to validate the
+// chart install on the cluster, like if a previous cilium install is not cleaned up properly
+// from the cluster.
```
What happens if it fails?

fristonio replied:
The caller has to make sure that there are no leftover cilium components in the cluster. I will add this line to the documentation of the function.

christarazi reviewed (commented):
LGTM! I left a few more comments mostly around documentation and making things consistent. The core of the work is good to go. :rocket:

christarazi reviewed (approved):
:tada:
